test/ui/LoginTestPage.ts

```diff
@@ -0,0 +1,100 @@
+    private errorMessage: By = By.id('error-message');
+
+    constructor(driver: WebDriver) {
+        this.driver = driver;
+    }
+
+    static async initialize(): Promise<LoginTestPage> {
+        const driver = await driverBuilder.driverBefore();
+        return new LoginTestPage(driver);
+    }
+    async open(url: string): Promise<void> {
+        await this.driver.get(url);
+    }
+
+    async enterUsername(username: string): Promise<void> {
+        const element = await this.driver.findElement(this.usernameField);
+        await element.sendKeys(username);
+    }
+
+    async enterPassword(password: string): Promise<void> {
+        const element = await this.driver.findElement(this.passwordField);
+        await element.sendKeys(password);
+    }
+
+    async clickLogin(): Promise<void> {
+        const element = await this.driver.findElement(this.loginButton);
+        await element.click();
+    }
+    async clickLogout(): Promise<void> {
+        const element = await this.driver.findElement(this.logoutButton);
+        await element.click();
+    }
+
+    async getLoggedToHomepageSuccessfully(): Promise<string> {
+        const element = await this.driver.findElement(this.loggedToHomepageSuccessfully);
+        return await element.getText();
+    }
+    async getloggedOutSuccessfully(): Promise<string> {
+        const element = await this.driver.findElement(this.loggedOutSuccessfully);
+        return await element.getText();
+    }
+
+    async getErrorMessageText(): Promise<string> {
+        const element = await this.driver.findElement(this.errorMessage);
+        return await element.getText();
+    }
+
+    async enterInvalidUsername(username: string): Promise<void> {
+        const element = await this.driver.findElement(this.usernameField);
+        await element.sendKeys(username);
+    }
+
+    async enterInvalidPassword(password: string): Promise<void> {
+        const element = await this.driver.findElement(this.passwordField);
+        await element.sendKeys(password);
+    }
+
+    async testLoginSuccess(username: string, password: string, expectedText: string): Promise<void> {
+        await this.open(LoginTestPage.URL);
+        await this.enterUsername(username);
+        await this.enterPassword(password);
+        await this.clickLogin();
+        const actualText = await this.getLoggedToHomepageSuccessfully();
+        expect(actualText).to.equal(expectedText);
+    }
+
+    async testLoginFailure(username: string, password: string, expectedError: string): Promise<void> {
+        await this.open(LoginTestPage.URL);
+        await this.enterInvalidUsername(username);
+        await this.enterInvalidPassword(password);
+        await this.clickLogin();
+        const errorMessage = await this.getErrorMessageText();
+        expect(errorMessage).to.equal(expectedError);
+    }
+    async testLogoutSuccess(username: string, password: string, expectedText: string): Promise<void> {
```

Comment:
Logout success does not need its own test. Add one like to logout in the testLoginSuccess. Alot of these tests are repetitive and unnecessary